### PR TITLE
Cleanup: Move array indexing validation out of apis package

### PR DESCRIPTION
**Pull request by lbernick**
We validate that indexing references to array parameters are in bounds,
based on the default values of the parameters and the parameter values
passed in from a PipelineRun/TaskRun. This validation happens in the reconciler,
since it requires comparing the Task/Pipeline spec with the parameters
defined in the TaskRun/PipelineRun.

Prior to this commit, validation code was in the apis package, even though it was called
in the reconciler. This is confusing, because this code cannot be used to validate a Task/Pipeline
spec in isolation; it can only be used in the context of TaskRuns and PipelineRuns.
This commit moves this validation logic to the reconciler and does not introduce any functional changes.

Related: https://github.com/tektoncd/pipeline/issues/6607

/kind cleanup

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- n/a Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- n/a Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- n/a Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/resources/validate_params.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/resources/validate_params.go:120]: I remember moving to api pkg as a member function is a suggestion from @pritidesai and thus introduced in https://github.com/tektoncd/pipeline/pull/6180. 
I wonder if @pritidesai has any comments here? 

I think if this ValidateParamArrayIndex function is not called in validation webhook, maybe we should avoid putting it in the `_validate.go` file? So I prefer this PR's change.

**Review by Yongxuanzhang — COMMENTED** [review]
Thanks! This is a huge improvement! 
I wonder should we remove this as well? https://github.com/tektoncd/pipeline/blob/09d422cff057f67170b4c2f76097ac6ffded33ef/pkg/reconciler/taskrun/resources/apply.go#L66

File: pkg/reconciler/taskrun/resources/taskref.go
Comment (lbernick) [reply pkg/reconciler/taskrun/resources/taskref.go:200]: I'd like to do this in a separate PR, since it would add a lot of new validation for remote tasks that we aren't already doing.

File: pkg/reconciler/pipelinerun/resources/validate_params.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/resources/validate_params.go:120]: I agree; I don't think this makes sense as a member function for pipelines since it really only makes sense in the context of a pipelinerun.

**Comment by lbernick** [comment]
Thanks @Yongxuanzhang I've removed the additional reconciler checks you linked.

File: pkg/reconciler/taskrun/resources/taskref.go
Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/taskref.go:200]: Ok got it!

**Comment by lbernick** [comment]
@tektoncd/core-maintainers please take a look, this is one of our remaining v1 blockers

File: pkg/apis/pipeline/v1beta1/param_types.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1beta1/param_types.go:124]: Do we need docstring for this since it is changed to public? 🤔 But why the lint is not failing maybe we don't need? Or do we need it to be public? I tried to search the references but only found it used in v1beta1, maybe I missed?

File: pkg/apis/pipeline/v1beta1/task_validation_test.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1beta1/task_validation_test.go:1757]: There are two cases name not updated: "steps reference invalid"->"steps reference"; "stepTemplate reference invalid" -> "stepTemplate reference"

File: pkg/reconciler/taskrun/resources/taskref.go
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskref.go:202]: Should we move "fmt.Errorf("failed to convert obj %s into Task", obj.GetObjectKind().GroupVersionKind().String())" into this function like readRuntimeObjectAsPipeline ?

**Review by Yongxuanzhang — COMMENTED** [review]
looks good to me! Just have some minor comments/questions. 
Let me know when we need a lgtm

**Comment by lbernick** [comment]
@Yongxuanzhang done, thanks!

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

**Review by jerop — CHANGES_REQUESTED** [review]
@lbernick do you mind breaking this down into smaller pull requests? it is difficult to review all the changes at once, I've been trying and it's taking a while -- a suggested breakdown:
1. fix indexing validation to check for beta ff instead of alpha ff
1. add array validation to when expressions
1. move validation from reconciler to webhook

**Comment by lbernick** [comment]
> @lbernick do you mind breaking this down into smaller pull requests? it is difficult to review all the changes at once, I've been trying and it's taking a while -- a suggested breakdown:
> 
> 1. fix indexing validation to check for beta ff instead of alpha ff
> 2. add array validation to when expressions
> 3. move validation from reconciler to webhook

Sure! Part 1 is already handled by @Yongxuanzhang's PR https://github.com/tektoncd/pipeline/pull/6613; I'll open a PR for part 2

**Comment by lbernick** [comment]
/hold
I've split off another refactoring PR: https://github.com/tektoncd/pipeline/pull/6642

File: pkg/reconciler/pipelinerun/resources/validate_params.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/resources/validate_params.go:151]: Sorry I know this may not be a change we need to have in this PR, but should we have more specific error message here? I don't remember why I use "non-existent param references", does it make more sense to say that those params are out of bound? If so I can fix this in another PR.

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

File: pkg/reconciler/pipelinerun/resources/validate_params.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/resources/validate_params.go:151]: If you wouldn't mind fixing this in a separate PR that would be great!

File: pkg/apis/pipeline/v1/task_validation.go
Comment (jerop) [inline pkg/apis/pipeline/v1/task_validation.go:635]: this is a bit different from the function it's replacing?

**Review by jerop — COMMENTED** [review]
@lbernick there's one more change happening in this PR around checking for feature flags, do you mind doing it separately? or update the commit message and PR description to include it? though it'd be cleaner to separate it out

thank you for the hard work in splitting up this PR so far!

**Comment by lbernick** [comment]
@jerop opened https://github.com/tektoncd/pipeline/pull/6646

File: pkg/apis/pipeline/v1/pipeline_validation.go
Comment (lbernick) [reply pkg/apis/pipeline/v1/pipeline_validation.go:694]: The logic is a bit different now.
Previously, it was "if we don't have beta fields enabled, don't validate array index bounds checking".
Now it is "if there is array indexing and we don't have beta fields enabled, return an error".

File: pkg/apis/pipeline/v1/task_validation.go
Comment (lbernick) [reply pkg/apis/pipeline/v1/task_validation.go:635]: Yes-- see the other comment I made; it's the same reason.

**Comment by lbernick** [comment]
@jerop @Yongxuanzhang this is ready for review-- it's now just the last components of cleanup/refactoring

File: pkg/reconciler/taskrun/resources/validate_params.go
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/validate_params.go:16]: Sorry I probably forgot to update this, do you mind updating it?

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

File: pkg/reconciler/taskrun/resources/validate_params.go
Comment (lbernick) [reply pkg/reconciler/taskrun/resources/validate_params.go]: done. I also moved some duplicated logic from pipelinerun/resources to taskrun/resources.

File: pkg/apis/pipeline/v1/param_types.go
Comment (dibyom) [inline pkg/apis/pipeline/v1/param_types.go:200]: since this is now exported, shouldn't it have its own unit tests?

File: pkg/apis/pipeline/v1/param_types.go
Comment (dibyom) [inline pkg/apis/pipeline/v1/param_types.go:237]: same as above

File: pkg/apis/pipeline/v1/param_types.go
Comment (dibyom) [reply pkg/apis/pipeline/v1/param_types.go:200]: alternatively, we can remove the receiver and turn this into a regular function and move it to the package its being used in?

Comment (lbernick) [reply pkg/apis/pipeline/v1/param_types.go:200]: added tests!

File: pkg/apis/pipeline/v1/param_types.go
Comment (lbernick) [reply pkg/apis/pipeline/v1/param_types.go:237]: added tests as well

**Pull request merged**
